### PR TITLE
Add HistogramPlot and minor fix to Plot

### PR DESCRIPTION
**Pull request by fvisin**
- Add HistogramPlot, an extension to plot histograms of the mean of each component of
a channel
- Call push only once at the beginning in Plot, as afterwards it is sufficient to modify the data source for the plot to update

### CODE REVIEW

File: blocks/extensions/plot.py [hunk 1]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
+        over the dataset.
+
+        .. warning::
+            This extension assumes that a bokeh server is running at the
+            provided url and won't try to start a server autonomously. To
+            start a bokeh server run `bokeh-server --ip=<ip> --port=<port>`.
+            If you want the server to accept connections from any address use
+            <ip>= 0.0.0.0.
+
+        Parameter
```
Comment (dwf) [inline blocks/extensions/plot.py]: Parameters*

File: blocks/extensions/plot.py [hunk 2]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
+        ---------
+        document : string
+            The name of the document.
+        channels : list
+            The channels to be displayed as a histogram.
+        server_url : string
+            The url of the server
```
Comment (dwf) [inline blocks/extensions/plot.py]: Period, and blank line before """

File: blocks/extensions/plot.py [hunk 3]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
+        self.session = Session(
+            root_url=self.server_url,
+            load_from_config=False)
+        output_server(self.document, session=self.session)
+
+
+class PlotHistogram(SimpleExtension):
+
+    def __init__(self, document, channels, server_url='default', **kwargs):
+        """ Live plot an histogram of mean and std dev of the provided channels
```
Comment (dwf) [inline blocks/extensions/plot.py]: No space after the """, and I think that pep257.py will complain if any of the docstring lines are >75 chars wide

File: blocks/extensions/plot.py [hunk 4]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
+        channels : list
+            The channels to be displayed as a histogram.
+        server_url : string
+            The url of the server
+        """
+        self.document = document
+        self.channels = channels
+        self.server_url = server_url
 
-    def __setstate__(self, state):
```
Comment (fvisin) [inline blocks/extensions/plot.py]: Is this ever called somewhere? I removed it as it would mess with the session of the plot. The same goes for `__getstate__`

File: blocks/extensions/plot.py [hunk 5]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
+                kl_std_dev = value.std(axis=0)
+                nhid = len(kl_mean)
+
+                fig = figure(title=key, plot_width=1000)
+                # display alternating color histograms
+                fig.quad(top=kl_mean[::2], bottom=0, left=range(nhid)[::2],
+                         right=range(nhid+1)[1::2], fill_color='#66E066',
+                         line_color="black")
+                fig.quad(top=kl_mean[1::2], bottom=0, left=range(nhid)[1::2],
+                         right=range(nhid+1)[2::2], fil_color='#66C2FF',
```
Comment (dwf) [inline blocks/extensions/plot.py]: I'm going to guess that `fil_color` is a typo for `fill_color`? Also you'll rpobably get some pep8 errors on this line

File: blocks/extensions/plot.py [hunk 6]
```diff
@@ -136,14 +132,86 @@ def preexec_fn():
...
-        state['sub'] = None
-        return state
+        self.session = Session(
+            root_url=self.server_url,
+            load_from_config=False)
+        output_server(self.document, session=self.session)
+
+
+class PlotHistogram(SimpleExtension):
+
```
Comment (dwf) [inline blocks/extensions/plot.py]: No need for a blank line here ;)

File: blocks/extensions/plot.py [hunk 7]
```diff
@@ -104,23 +98,25 @@ def do(self, which_callback, *args):
         log = self.main_loop.log
         iteration = log.status['iterations_done']
         i = 0
-        for key, value in log.current_row.items():
```
Comment (dwf) [inline blocks/extensions/plot.py]: I don't really see why this change is necessary?

File: blocks/extensions/plot.py (hunk 4 above)
Comment (dwf) [inline blocks/extensions/plot.py]: These are used by pickle for serialization/deserialization.

https://docs.python.org/2/library/pickle.html#object.__getstate__

File: blocks/extensions/plot.py (hunk 5 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: No pep8 (at least according to flake8) and...yes, it's a typo but it was working anyway, apparently. Bokeh magic I guess :) Anyway I'll fix it!

**Comment by fvisin** [comment]
Wow, @dwf I never received a review so fast! I am working on it.

File: blocks/extensions/plot.py (hunk 4 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: I see. In this case I will restore them and also implement the equivalent for `HistogramPlot` too.

File: blocks/extensions/plot.py (hunk 7 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: My fault, it's a leftover of a modification that I have then deleted. I will restore it as it was.

File: blocks/extensions/plot.py (hunk 3 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: Apparently pep257 is not complaining for the length of the docstring lines (running it locally). I will check Travis output when I am done with the rest.

File: blocks/extensions/plot.py [hunk 8]
```diff
@@ -4,14 +4,12 @@
 from subprocess import Popen, PIPE
 
 try:
-    from bokeh.plotting import (curdoc, cursession, figure, output_server,
```
Comment (dwf) [inline blocks/extensions/plot.py]: were curdoc and cursession never used, I guess?

Comment (fvisin) [inline blocks/extensions/plot.py]: curdoc was used by `__setstate__`, while cursession() is now unused because I use self.session. This prevents problems when you have multiple extensions plotting (i.e. multiple sessions open)

**Comment by fvisin** [comment]
Travis error doesn't seem related to my PR and I should have addressed all your requests.

**Comment by dwf** [comment]
I restarted Travis but these did in fact look related... the demos run with
the plotting extension, and there was an error regarding Bokeh schema.

File: blocks/extensions/plot.py [hunk 9]
```diff
@@ -63,9 +61,7 @@ class Plot(SimpleExtension):
     server_url : str, optional
         Url of the bokeh-server. Ex: when starting the bokeh-server with
         ``bokeh-server --ip 0.0.0.0`` at ``alice``, server_url should be
-        ``http://alice:5006``. When not specified the default configured
-        by ``bokeh_server`` in ``.blocksrc`` will be used. Defaults to
-        ``http://localhost:5006/``.
```
Comment (bartvm) [inline blocks/extensions/plot.py]: Why did you remove this configuration setting?

File: blocks/extensions/plot.py [hunk 10]
```diff
@@ -136,14 +130,103 @@ def preexec_fn():
...
+
+        This extension live plots an histogram of the mean value of each of
+        the provided channels over the monitored dataset. It will also
+        display confidence intervals representing the standard deviation of
+        the channel over the dataset.
+
+        .. warning::
+            This extension assumes that a bokeh server is running at the
+            provided url and won't try to start a server autonomously. To
+            start a bokeh server: `bokeh-server --ip=<ip> --port=<port>`.
```
Comment (bartvm) [inline blocks/extensions/plot.py]: ``

File: blocks/extensions/plot.py [hunk 11]
```diff
@@ -136,14 +130,103 @@ def preexec_fn():
...
+            If you want the server to accept connections from any address
+            use <ip>= 0.0.0.0.
```
Comment (bartvm) [inline blocks/extensions/plot.py]: ``

File: blocks/extensions/plot.py (hunk 9 above)
Comment (dwf) [inline blocks/extensions/plot.py]: Indeed, this documentation looks useful.

File: blocks/extensions/plot.py [hunk 12]
```diff
@@ -136,14 +130,103 @@ def preexec_fn():
...
+                # position relatively to the previous histogram
+                std_hist_bottom = kl_mean - kl_std_dev/2.0
+                std_hist_top = kl_mean + kl_std_dev/2.0
+                fig.quad(top=std_hist_top, bottom=std_hist_bottom,
+                         left=[el + 0.47 for el in range(nhid)],
+                         right=[el + 0.53 for el in range(nhid)],
+                         fill_color='black', line_color="black")
+        push()
+
     def __setstate__(self, state):
```
Comment (bartvm) [inline blocks/extensions/plot.py]: Seems like me that you can get rid of `__getstate__` and `__setstate__` entirely here. `sub` used to contain the subprocess, but you don't have it here, so the default behavior of `pickle` should work.

File: blocks/extensions/plot.py [hunk 13]
```diff
@@ -136,14 +130,103 @@ def preexec_fn():
             logger.info('Plotting server PID: {}'.format(self.sub.pid))
         else:
             self.sub = None
-        output_server(self.document, url=self.server_url)
+
+        self.session = Session(
+            root_url=self.server_url,
```
Comment (bartvm) [inline blocks/extensions/plot.py]: Judging from the errors on Travis, it's this line that is causing problems. If it's not for you, that might be because of different Bokeh versions?

File: blocks/extensions/plot.py (hunk 9 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: Apparently I failed to merge https://github.com/bartvm/blocks/commit/9c8ad24f9d728f1cc534c6fc0d9cd7a6a2991d24 for some reason.

File: blocks/extensions/plot.py (hunk 13 above)
Comment (fvisin) [inline blocks/extensions/plot.py]: I just checked and on my computer this line is not causing problems. I am using the last version of bokeh, that was released a few days ago.

Comment (fvisin) [inline blocks/extensions/plot.py]: Where do you see the Travis errors? The errors I see are two: a coverage decreased error and an error from Scrutinizer that seems related to another file: `blocks/bricks/conv.py:238:1: W293 blank line contains whitespace`. Travis build apparently passed.

Comment (bartvm) [inline blocks/extensions/plot.py]: They disappeared when you merged the `config` commit. The problem was there.

Comment (fvisin) [inline blocks/extensions/plot.py]: Oh well, great! ;)

File: blocks/extensions/plot.py [hunk 14]
```diff
@@ -80,52 +79,76 @@ def __init__(self, document, channels, open_browser=False,
...
-        self.plots = {}
+        self.data_sources = {}
         self.start_server = start_server
         self.document = document
         self.server_url = server_url
+
+        kwargs.setdefault('after_epoch', True)
+        kwargs.setdefault("before_first_epoch", True)
+
+        # set x_axis_label to the most frequent option
```
Comment (fvisin) [inline blocks/extensions/plot.py]: This is not very elegant, but I didn't want to add parameters to the constructor. Basically I try to infer if the x-axis should be over iterations or batches, instead of just assuming that iteration is ok.

**Comment by fvisin** [comment]
I pushed another minor commit. The changes are mostly aesthetic, but they improve the plot quite a bit in my opinion. If you don't have any other comment, I don't plan to modify it further.

**Comment by fvisin** [comment]
Travis build passed, is it ready for merge or I am not noticing something?

**Comment by fvisin** [comment]
Hey, is there a reason why this PR has not been merged yet? If there is something missing can you please point that out?

**Comment by bartvm** [comment]
Personally I'd like to wait till we have `blocks-contrib`/`blocks-extra` because it's a lot of code that I wouldn't consider essential to teh core. I'm sorry this has taken so long, we had exams last week and I've simply been too busy to get that sorted out. If people feel there's a need to merge this sooner I'm happy to hear it.

**Comment by fvisin** [comment]
Don't worry! I know you are busy, I just wanted to make sure it wasn't due to something I was missing.

For what concerns the merge, I would like to see it merged so that I can use it without copying and pasting the code from/to different branches, but it's just a bit annoying, I wouldn't define it a big pain. Nonetheless, I fear that the creation of `blocks-contrib`/`blocks-extra` will be continuously delayed, so.......maybe you could merge it, unless you actually expect the auxiliary repos to be online soon? :)

**Comment by bartvm** [comment]
I'll try to do my best to get it up and running soon! So let's hope it won't end up being continuously delayed. The annoying thing about merging it is that it then becomes a copy-paste job later to move things later, breaking people's code, etc.

There's no need to copy-paste by the way. Whenever you want to use this, you can just branch `master` and merge `plot_histogram` (plus whatever other branches you need). I use that workflow quite a lot, because I've got different Blocks branches with functionality all over the place.

**Comment by rizar** [comment]
@fvisin , seems like you also have batch normalization PR in this one. Also I am afraid that your extension is not training resumption friendly, see `__setstate__` and `__getstate__` methods of the Plot extension.

Regarding the merge, I think I support @bartvm in his reluctance. For every bit of code merged in Blocks we become responsible, we have to review further PRs and bug fixes. We would much rather concentrate on the core functionality than doing that.

Cross-post!

**Comment by rizar** [comment]
In fact I think that creation of a non-tested repo that would simply contain bits and pieces of Blocks-compatible code is a matter of 10 minutes, and I can gladly do that. All the rest, like `setup.py` and other things, can be postponed: people can use `PYTHONPATH` to use this repo for beginning.

@bartvm, do you have anything against going this way?

**Comment by bartvm** [comment]
It all depends on what you want to do. If it's just a collection of code snippets, I don't think it should even go in a repo, in that case we can just add a collection of links to the wiki of this repo. If we want to set up Travis CI, Scrutinizer, write documentation, move code from here to there, etc. it becomes more than 10 minutes.

Either way, the main thing holding it back is the fact that we're supposed to create the repos as part of the [new GitHub organisation](https://github.com/mila-udem), but I don't have admin rights there, so I've been unable to do anything :p I'll poke Fred/Pascal about that!

**Comment by rizar** [comment]
Collections of links does not do the job because one can not check out them all and import from this the checkout directory.

The whole transition is a of course a lot of work, especially moving things out from Blocks, but now I am more worried about making it possible for people to share their useful code in a convenient way.

**Comment by fvisin** [comment]
@bartvm thank you for the merge trick. I didn't think about it, very useful!

@rizar You are right, I was using batch normalization for a test and forgot to delete the commit. I did it now. For what concerns training resumption, Bart and I agreed on removing that part: 
> sub used to contain the subprocess, but you don't have it here, so the default behavior of pickle should work.

For what concerns the PR, it can wait until the extras repo will be available.

**Comment by rizar** [comment]
@fvisin , for your information, [blocks-extras](https://github.com/mila-udem/blocks-extras) is already there!
